scio-tensorflow/src/main/scala/com/spotify/scio/tensorflow/TensorFlowDoFns.scala

```diff
@@ -105,48 +104,92 @@ private[tensorflow] abstract class SavedBundlePredictDoFn[T, V](
 object SavedBundlePredictDoFn {
   def forInput[T, V](
     uri: String,
+    fetchOps: Option[Seq[String]],
     options: TensorFlowModel.Options,
-    fetchOp: Seq[String],
+    signatureName: String,
     inFn: T => Map[String, Tensor[_]],
     outFn: (T, Map[String, Tensor[_]]) => V
-  ): SavedBundlePredictDoFn[T, V] = new SavedBundlePredictDoFn[T, V](uri, options) {
-    override def extractInput(input: T): Map[String, Tensor[_]] = inFn(input)
+  ): SavedBundlePredictDoFn[T, V] = new SavedBundlePredictDoFn[T, V](uri, signatureName, options) {
+    var exportedFetchOps: Map[String, String] = _
+
+    private def requestedFetchOps: Map[String, String] = {
+      fetchOps
+        .map { tensorIds =>
+          tensorIds.map { tensorId =>
+            tensorId -> exportedFetchOps(tensorId)
+          }.toMap
+        }
+        .getOrElse(exportedFetchOps)
+    }
+
+    override def createResource(): TensorFlowModel = {
+      val model = TensorFlowLoader
+        .create(Id.create(id), uri, options, signatureName)
+        .get(Duration.ofDays(Integer.MAX_VALUE))
+      // by doing it here we make sure we only do it once
+      exportedFetchOps = model.outputsNameMap().asScala.toMap
 
-    override def extractOutput(input: T, out: Map[String, Tensor[_]]): V = outFn(input, out)
+      model
+    }
 
-    override def outputTensorNames: Seq[String] = fetchOp
+    override def extractInput(input: T): Map[String, Tensor[_]] = {
+      val extractedInput = inFn(input)
+      extractedInput.map {
+        case (tensorId, tensor) =>
+          getResource.inputsNameMap().get(tensorId) -> tensor
+      }.toMap
+    }
 
+    override def extractOutput(input: T, out: Map[String, Tensor[_]]): V =
+      outFn(input, requestedFetchOps.mapValues(out(_)))
+
+    override def outputTensorNames: Seq[String] = requestedFetchOps.values.toSeq
   }
 
+  /**
+   * Note: if fetchOps isn't provided, then all outputs defined in the signature of the model
+   * are retrieved. This can be expensive and unwanted, depending on the model.
+   */
   def forTensorFlowExample[T, V](
     uri: String,
     exampleTensorName: String,
-    signatureName: String,
+    fetchOps: Option[Seq[String]],
     options: TensorFlowModel.Options,
+    signatureName: String,
     outFn: (T, Map[String, Tensor[_]]) => V
   )(implicit ev: T <:< Example): SavedBundlePredictDoFn[T, V] =
-    new SavedBundlePredictDoFn[T, V](uri, options) {
-
-      var fetchOps: Map[String, String] = _
+    new SavedBundlePredictDoFn[T, V](uri, signatureName, options) {
+      var exportedFetchOps: Map[String, String] = _
+
+      /* tensorId refers to the public/export name of the tensor */
+      private def requestedFetchOps: Map[String, String] = {
```

Comment:
this will be executed for every element. probably a good idea to do it only on setup. The `exportedFetchOps ` are not really used outside of this function, so maybe keep a reference to this map instead of that one?